### ISSUE TITLE
AEH-W4A1

### ISSUE THREAD

**Issue by Maxetto** [deiger/AirCon#1]
Hi,
I've found your repository the first day of creation. I've tried the script with AEH-W4A1 but this module does not seem to have an open LAN port.
It has 8888 TCP open but it just replays everything you send to it.
With Wireshark I found it communicates to the servers with port 1024 TCP, but it is closed/filtered on LAN.

It communicates with a couple of servers, but the main one is:
hi-smartlife.topfuturesz.com (TCP Port 7820)

I've been able to replay some TCP packets to turn on/off, change the fan speed, get and set the temperature, turn off the dimmer and other things.

I will try the new script, but upon sniffing traffic, I couldn't find any lan_key in the communication between the servers, so I think this module is out of luck with LAN control.

**Comment by deiger** [comment]
Hi,
Which app are you using?
I haven't tried https://play.google.com/store/apps/details?id=com.qd.android.livehome (Hi-Smart Life), but it should work for https://play.google.com/store/apps/details?id=com.accontrol.europe.hisense (Hi-Smart AC).
I'll try finding out if the former is using the same protocol.

**Comment by deiger** [comment]
I pried into the Hi-Smart Life app, and it seems entirely different. It is not based on aylanetworks code, but has something by xinlianfeng.
I'm afraid if none of the aylanetworks-based apps works, then a different solution would be needed.

**Comment by Maxetto** [comment]
https://github.com/cgdgithub/ControlJavaEdition

Searching for xinlianfeng I found this repository, it seems that this is the same way they communicate with Hi-Smart Life app, it also states that the module port is 8888 which is correct, and also seems that this port is used to receive authenticated (only with username) mqtt commands.

**Comment by deiger** [comment]
Cool!
I couldn't find any mention for MQTT within the current app, but it is definitely worth a shot. 
If you could configure it to work against an MQTT broker your life would be much easier.

**Comment by bannhead** [comment]
Hi guys,
I'm working on reverse engineering AEH-W4A1. Luckily it is much simpler than newer revision: if on lan, after contacting the china server to obtain the list of ac, it communicates with them directly on port 8888 via a plain serial protocol.

I used Packet Capture, an android app to MITM sniff packets.

For example, if you raw send **AT+XMV** to port 8888 of ac you obtain a response like **+XMV:4.4.6** (it can be useful to identify the AEH-W4A1 module and also to search in the LAN of all the conditioners)

And if you send 
**F4F500400C00000101FE0100006600000001B3F4FB**
the response is the current state, for example
**F4F50140490100FE01010101006600010000201A211D80808001A10000000000000000000000050000000000215E2300000000000000000000000000080000000000000000000000000000000000053DF4FB**

I searched a lot online information about the protol used, but I found only an old pamplet about a similar serial protol developed together by Hisense and Hitachi. Probably an archaic version of the current one, but of no use.

Unfortunately I do not have the skills to develop an add-on for Home Assistant, but if someone wanted to do it I could put myself to decode all the bytes of the protocol and reconstruct its functioning. I have already done some work ;)

**Comment by TheFlatline** [comment]
Hi!
I'd be terribly interested in the decoding of the protocol! Once it is done (i.e. we need just to send bytes to device_ip:8888 to get status and send commands I think it should be relatively trivial to implement control via home assistant!

**Comment by Maxetto** [comment]
@bannhead Unfortunately with +XMV:4.4.7, I only get a response directly from the AC if i send AT+XMV.
All the other commands it just echoes what you wrote.

The protocol seems to be a somehow encoded JSON string, you can learn more here: https://github.com/cgdgithub/ControlJavaEdition/blob/master/src/com/xinlianfeng/android/livehome/devices/base/DevicesLogic.java#L461

Anyway, you can do guessing with many replies, for example, in the status response **F4F50140490100FE01010101006600010000201A211D80808001A10000000000000000000000050000000000215E2300000000000000000000000000080000000000000000000000000000000000053DF4FB**

in the series of bytes **201A21** the byte **20** means that the AC is Off, while **28** means that the AC is On; while the byte **21** represents the current temperature (in my testing, **1C** was 28°C, **1B** 27°C, **1A** 26°C, **19** 25°C and **17** 23°C)

**Comment by deiger** [comment]
Unfortunately the actual parsing logic here is hidden in the JNI-wrapped libhisense.so file, in functions like Java_com_xinlianfeng_android_livehome_hisense_ProtocolNative_parsePara and Java_com_xinlianfeng_android_livehome_hisense_ProtocolNative_buildPara.
Likely the best option to analyse the entire protocol would be to link to the .so file in AndroidStudio and feed it with different inputs.

**Comment by bannhead** [comment]
It's a serial protocol, JSON is used only in app for code semplicity; communication to and from AC are raw bits sent in tcp over port 8888.
I extracted deviceprotocol.java and obtained the bit code to create and read data packets.

Standard packet
`F4F50140490100FE010101010066000108002A1D1F1B80808001D10000000000000000C000800500000000001E481B00000000000000000000000000080000000000000000000000000000000000069DF4FB`
F4F5 header
01 received packet (00 to send)
F4FB footer
40 need further analysis - don't change
49 need further analysis
0100FE0101010100 need further analysis - don't change
6600 packet type (66 ->102) and sub type
010 need further analysis - don't change
8002A1D1F1B80808001D10000000000000000C000800500000000001E481B00000000000000000000000000080000000000000000000000000 data
00000000006 need further analysis
9D 1 byte checksum on data
F4FB footer

And here the data part converted in binary and divided using the table obtained from java analysis:
` 1 0000000 000000 1 (5->) 0 1010 00011101 00011111 00011011 (10->) 10000000 10000000 10000000 0 00000 (15->) 0 1110 100010 0 0 (20->) 00000 0 000000 0 0 (25->) 00000 0 000000 0 0 (30->) 00000 0 000000 000 000 (35->) 0 0 0 0 0 (40->) 0 0 0 0 0 (45->) 0 0 0 0 0 (50->) 1 1 0 0 0 (55->) 0 0 0 0 0 (60->) 0 0 0 00010000 00000000 (65->) 10100000 00000000 00000000 00000000 00000000 (70->) 00000011 11001001 00000011 0110000000000000 0000000000000000 (75->) 00000000 00000000 00000000 00000000 0000000000000000 (80->) 00000000 00 0 0 0 (85->) 0 0 0 0 0 (90->) 0 0 0 0 0 (95->) 0 0 0 0 00001000 (100->) 00000000 00000000 00000000 00000000 00000000 (105->) 00000000 00000000 00000000 00000000 00000000 (110->) 0000 0 00000000 0000000`

`1   WindSpeedMode
2   AirVolume
3   SleepMode
4   WindDirectionSwitch
5   LaunchSwitch
6   WorkMode
7   IndoorSetTemp
8   IndoorCurrentTemp
9   IndoorPipeTemp
10  IndoorSetHumi
11  IndoorCurrentHumi
12  RecvFeelIndoorTemp
13  FeelControl
14  FeelIndoorTempOffset
15  TempDisplayMode
16  AutoAndDehumiModeTempOffset
17  NormalTimingValue
18  NormalTimingValidity
19  RTCHourExplain
20  RTCHourValue
21  RTCMinuteExplain
22  RTCMinuteValue
23  RTCPowerOnHourExplain
24  RTCPowerOnControl
25  RTCPowerOnHourValue
26  RTCPowerOnMinuteExplain
27  RTCPowerOnMinuteValue
28  RTCPowerOffHourExplain
29  RTCPowerOffControl
30  RTCPowerOffHourValue
31  RTCPowerOffMinuteExplain
32  RTCPowerOffMinuteValue
33  WindValvePosition
34  DehumiMode
35  DualModeSwitch
36  StrongSwitch
37  CombinConserveEnergySwitch
38  ConserveEnergySwitch
39  ElectricHeatSwitch
40  NaturalWindSwitch
41  LeftRightWindSwitch
42  UpDownWindSwitch
43  SmokeSwitch
44  VoiceControl
45  MuteSwitch
46  SmartEyeSwitch
47  OutdoorCleanSwitch
48  IndoorCleanSwitch
49  FanSwitch
50  CleanerSwitch
51  IndoorElectricityBoard
52  RightWindSwingSwitch
53  LeftWindSwingSwitch
54  IndoorStrainerCleanStatus
55  TempDisplaySwitch
56  LedSwitch
57  DisplayScreenShineSwitch
58  BackgroundLightSwitch
59  IndoorEEPROMUpgrade
60  Model
61  BeforeWifiControl
62  BeforeIrAndButtonControl
63  IndoorAlarm1
64  IndoorAlarm2
65  CompressorRunHz
66  CompressorTargetHz
67  ToDriverHz
68  OutdoorEnvironmentTemp
69  OutdoorCondenserTemp
70  CompressorExhaustTemp
71  TargetExhaustTemp
72  OutdoorEXVOpening
73  Uab
74  Ubc
75  Uca
76  Iab
77  Ibc
78  Ica
79  UDCBus
80  Iuv
81  FanRunStatus
82  OutdoorUnitCurrentWorkStatus
83  FourWayValveStatus
84  OutdoorDefrostingCream
85  OutdoorFrost
86  DehumiValve
87  MultiSplit
88  TempControlPowerOff
89  ForceInnerStop
90  ForceInnerSpeed
91  ForceInnerWindValvePosition
92  FillGasIncreaseHan
93  CompressorPreheat
94  CompressorRibbonHeater
95  OutdoorElectricityBoard
96  OutdoorEEPROMUpgrade
97  OutdoorFaultDisplay
98  OilReturn
99  OutdoorAlarm1
100 OutdoorAlarm2
101 OutdoorAlarm3
102 OutdoorAlarm4
103 OutdoorAlarm5
104 OutdoorAlarm6
105 OutdoorAlarm7
106 OutdoorAlarm8
107 OutdoorAlarm9
108 IndoorFanRPM
109 OutdoorFanRPM
110 PM2.5Level
111 WhetherPM2.5
112 PM2.5%
113 DisplayScreenBrightness`

**Comment by Maxetto** [comment]
![image](https://user-images.githubusercontent.com/1360752/66150693-fc1d6f80-e615-11e9-8116-078298224068.png)

This is the output of my netcat on Windows with XMV 4.4.7, it replies to XMV request but not to the hex string.

**Comment by bannhead** [comment]
You must send not plain text, but a real bit string. You can create it using a hex editor like HxD and pass it to nc using < and > to redirect output to a file

**Comment by Maxetto** [comment]
Oh, it works. I don't get why sending plain HEX to the Hisense server works but not directly to the AC :P

Well this is a nice finding, LAN is Open \😃/

**Comment by deiger** [comment]
It's surprising that in this protocol they use Celsius, in the AEH-W4B1 protocol they rely on Fahrenheit.
If you get the entire protocol working I may be able to integrate it as another option to my code. I'd need volunteers for repeated testing though, as I don't have the relevant module.

**Comment by bannhead** [comment]
I have almost finished analysis of responses for packet type 102. This is the most important type because it reports the full state of the AC. Other packet types (3, 3 sub 1, 7 sub 1, 10 sub 4, 30, 101, 101 sub 32 and 202) are related to other things like setting modes (101), wifi (30) and so on.

[These](https://drive.google.com/file/d/1K6qdijNnSLKMMEIREe5DxBzMbA-cMeMT/view?usp=sharing
https://drive.google.com/file/d/1X0dlAzwSA3_JsJJbhk3O-2g2pSHQ5d1F/view?usp=sharing
https://drive.google.com/file/d/1mTMuDkfRTzD2l_iuy52PMcK0IvrfzF46/view?usp=sharing) are the links to download the case study and related tables to interpret fields and values.

**Comment by bannhead** [comment]
> It's surprising that in this protocol they use Celsius, in the AEH-W4B1 protocol they rely on Fahrenheit.
> If you get the entire protocol working I may be able to integrate it as another option to my code. I'd need volunteers for repeated testing though, as I don't have the relevant module.

It is possible to switch easily to °F and the other way around: all the logic start to use and report temperature consequently.

[This](https://github.com/passerby012745/device_aircondition/blob/v3/src/com/szsbay/livehome/openlife/aircondition/DeviceProtocol.java) is the source of my knowledge of the protocol.

On the DeviceProtocol and OffsetAttribute declarations explains the parameters for the creation and interpretation of the data field of the packets, with a few small variations that I indicated in my analysis, perhaps due to a revision of the protocol.

But the best thing is that most command packets can be written statically, except for those where a timer is set and little else!

If you tell me what else you may need I am at your disposal! I can create a list of commands to be sent to set the various parameters and obviously I am available for any clarification on the interpretation of the reply packets with the conditioner status.

**Comment by deiger** [comment]
In AEH-W4B1 everything is Fahrenheit, there is a parameter for changing the mode to Celsius, but it only affects the display on the AC.

Here is what I need:
- A list of parameters (preferably mapped to the parameter names [here](https://github.com/deiger/AirCon)). For each one the list of possible values, and whether the parameter is read-only.
- How to fetch the current status of a single/all parameters.
- How to update each parameter. A static string would be great.
- Is there a way to get an update once the AC's status is changed?

**Comment by bannhead** [comment]
I have done my dirty job!

> * A list of parameters (preferably mapped to the parameter names [here](https://github.com/deiger/AirCon)). For each one the list of possible values, and whether the parameter is read-only.

[List of parameters](https://drive.google.com/file/d/17tFmjm2SnLFfbvuwqV6vOdIoSnBR5Wsr/view?usp=sharing)

The read-only flag is only for JSON, because for AEH-W4A1 you obtain a response packet, as follow

> * How to fetch the current status of a single/all parameters.

[Response packet analysys](https://drive.google.com/file/d/18e6AR807B2pI3NyrdU8wCpqOinjLvnm5/view?usp=sharing)

Every packet contains all the values

> * How to update each parameter. A static string would be great.

[static command packets](https://drive.google.com/file/d/1NR-7vq-XIbPJSeQS_liNtLU3NyQSYWSs/view?usp=sharing)

Only the packets to set the on and off timers, which need to be created dynamically, are missing. If it can be useful to implement it, I already have the instructions to create the packets ;)

> * Is there a way to get an update once the AC's status is changed?

No, the Android app requires status every time it sends a command and when it is inactive about every 40 seconds.

I hope everything is clear and that you can add support to this device too

**Comment by bannhead** [comment]
For completeness, I point out that the protocol extrapolated [here](https://github.com/passerby012745/device_aircondition/blob/v3/src/com/szsbay/livehome/openlife/aircondition/DeviceProtocol.java) was the basis of everything, but that it is a non-exact correspondence with the current one. [This](https://drive.google.com/open?id=1mTMuDkfRTzD2l_iuy52PMcK0IvrfzF46) is the file with the original version and [this](https://drive.google.com/open?id=1uZ_sQSYu66WMk8VIos8s09U7xpdxClwD) one with the revision I made.

**Comment by deiger** [comment]
Thanks, I'll start merging it into the script.

**Comment by Maxetto** [comment]
Hi-Smart Life app updated today to version 5.2.9 with changelog "Updated the application protocol".
Did someone test the new version? What is changed and if it breaks our control method?

**Comment by deiger** [comment]
There are significant code changes, it's hard to exactly as the code when through some sort of obfuscation processing.
OTOH, I'd be surprised if the protocol itself has significantly changed, since that would require an update to the server code running on the AC as well (and they would also still need to handle the previous version for ACs that have not been updated).

**Comment by bannhead** [comment]
Ok, in my spare time I managed to do a working HA integration for the AEH-W4A1 module; post on HA forum [here](https://community.home-assistant.io/t/working-on-integration-for-hisense-aeh-w4a1-module/146243)

I ask those who have a different module like AEH-W4B1 and AEH-W4E1 if they can test the [library](https://pypi.org/project/pyaehw4a1/) on the local network to find out if they are compatible.

Those who own an AEH-W4A1 can copy the integration into custom components to try it while waiting for the pull request

**Comment by deiger** [comment]
@bannhead, Nice! The protocol for AEH-W4B1 is entirely different, no chance for it to work.

**Comment by happylichuanxi** [comment]
yeah,it seem that app or module version differs,the protocol differs,so it's a hard way.I have tried the serial comm to control it,using esp8266 and ttl-to-rs485 module,and it's a little difficult to analysis the code,any help would be grateful!

**Comment by happylichuanxi** [comment]
this is the code i have captured and i analysisd some of them:

AA 2A AC 00 00 00 10 00 00 02 01 04 00 48 35 00 00 00 04 04 00 00 04 01 04 01 0E 00 00 00 00 80 00 00 00 00 00 00 00 00 00 DE 18  //set_temp 26 cool
Temp,    //15 byte  (x*2-1)
mode(cool:28;auto:48;dry:38;fan:08;heat:18),    //14byte
swing(00,02,04,06,08,0A,0C,0E),    //31byte
fan_speed(03:speed_1,04:speed_2,05:auto,06:speed_3,08:speed_4)  //12byte
the rest i haven't figured out.

**Comment by bannhead** [comment]
@happylichuanxi could you grab various packets (switching modes, hor and vert swing on and off, set temp from min to max step by step, and everything else you can change and grab via serial bus) and send them to me?

**Comment by happylichuanxi** [comment]
> @happylichuanxi could you grab various packets (switching modes, hor and vert swing on and off, set temp from min to max step by step, and everything else you can change and grab via serial bus) and send them to me?
> 
> Can you indicate wifi module and ac brand&model?



> @happylichuanxi could you grab various packets (switching modes, hor and vert swing on and off, set temp from min to max step by step, and everything else you can change and grab via serial bus) and send them to me?
> 
> Can you indicate wifi module and ac brand&model?

sure,but it takes time, I will grab as soon as I can.

**Comment by abmantis** [comment]
@happylichuanxi were you able to get the ESP8266 to control the AC directly without using the hisense module?

**Comment by straga** [comment]
I am trying to connect directly to AC over RS485. esp32+uartRS485 = Works.
https://github.com/bannhead/pyaehw4a1/issues/1 - my test.


```
# Command: package len 240 bit

61: always 1

# Power
ON:
11 - Command: last always 1
10 - Status = 10

OFF:
01 - Command: last always 1
00 - Status = 00

len: 1

ON / OFF
						OFF: offset:    [20, 21, 61]
00000000000000000000 11 00000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
						OFF: offset:        [21, 61]
00000000000000000000 01 00000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

# Wind
Super Low Wind:
00001011 - Command: last always 1.
00001010 - Status = 10

Super High Wind
00010011 - Command:
00010010 - Status = 18

         Auto in AC mode only: offset:             [7, 61]
00000001 0000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
         Super Low Wind - offset:            [4, 6, 7, 61]
00001011 0000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
         Low Wind - offset:                  [4, 5, 7, 61]
00001101 0000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
         Middle Wind - offset:            [4, 5, 6, 7, 61]
00001111 0000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 
         High Wind - offset:                    [3, 7, 61]
00010001 0000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
         Super High Wind - offset:           [3, 6, 7, 61]
00010011 0000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

**Comment by miky2k** [comment]
I will try to use RS485+esp8266(esp-link)   can you give me baudrate  of rs485 ?

**Comment by straga** [comment]
baudrate=9600

**Comment by miky2k** [comment]
![IMG_20221012_174935](https://user-images.githubusercontent.com/6616232/195414031-749a8b77-5935-454c-88bc-c1f3796abd76.jpg)

I will use this adapter, only understand better , i send 30bytes @9600 baud with commands and i receive 30bytes ? Or i receive only one byte with status?

**Comment by straga** [comment]
all byte.

**Comment by 9600** [comment]
@miky2k Hello, yes, you called?

**Comment by miky2k** [comment]
I'm testing hardware with simple c code:
1) ESP8266 (esp-link firmware )
2) from linux i use  socat to create virtual serial port:  /usr/bin/socat -T 20 pty,link=/tmp/ttyS10 tcp:192.168.120.140:23
3) ``/*
 ============================================================================
 Name        : hisense8266.c
 Author      : far5893
 Version     :
 Copyright   : Your copyright notice
 Description : Hisense serial test
 ============================================================================
 */
#define TERMINAL    "/tmp/ttyS10"

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <termios.h>

int set_interface_attribs(int fd, int speed)
{
    struct termios tty;

    if (tcgetattr(fd, &tty) < 0) {
        printf("Error from tcgetattr: %s\n", strerror(errno));
        return -1;
    }

    cfsetospeed(&tty, (speed_t)speed);
    cfsetispeed(&tty, (speed_t)speed);

    tty.c_cflag |= (CLOCAL | CREAD);    /* ignore modem controls */
    tty.c_cflag &= ~CSIZE;
    tty.c_cflag |= CS8;         /* 8-bit characters */
    tty.c_cflag &= ~PARENB;     /* no parity bit */
    tty.c_cflag &= ~CSTOPB;     /* only need 1 stop bit */
    tty.c_cflag &= ~CRTSCTS;    /* no hardware flowcontrol */

    /* setup for non-canonical mode */
    tty.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP | INLCR | IGNCR | ICRNL | IXON);
    tty.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
    tty.c_oflag &= ~OPOST;

    /* fetch bytes as they become available */
    tty.c_cc[VMIN] = 1;
    tty.c_cc[VTIME] = 1;

    if (tcsetattr(fd, TCSANOW, &tty) != 0) {
        printf("Error from tcsetattr: %s\n", strerror(errno));
        return -1;
    }
    return 0;
}
uint8_t cmdOFF[]="000000000000000000000100000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000";

void cmd2bin(uint8_t * cmdi,uint8_t *cmdo){
	uint8_t b;
	for(int i=0;i<30;i++){
       b=0;
       b=(cmdi[i*8] == '0' ? 0  : 128);
       b+=(cmdi[i*8+1] == '0' ? 0 : 64);
       b+=(cmdi[i*8+2] == '0' ? 0 : 32);
       b+=(cmdi[i*8+3] == '0' ? 0 : 16);
       b+=(cmdi[i*8+4] == '0' ? 0 : 8);
       b+=(cmdi[i*8+5] == '0' ? 0 : 4);
       b+=(cmdi[i*8+6] == '0' ? 0 : 2);
       b+=(cmdi[i*8+7] == '0' ? 0 : 1);

	   cmdo[i]=b;
	}
}

int main(void) {
	char *portname = TERMINAL;
	int fd;
	int wlen;
	uint8_t c[30];
	cmd2bin(cmdOFF,c);
	fd = open(portname, O_RDWR | O_NOCTTY | O_SYNC);
	    if (fd < 0) {
	        printf("Error opening %s: %s\n", portname, strerror(errno));
	        return -1;
	    }
	    /*baudrate 115200, 8 bits, no parity, 1 stop bit */
	    set_interface_attribs(fd, B9600);
	    //set_mincount(fd, 0);                /* set to pure timed read */
	    wlen = write(fd,c,30);
	       if (wlen != 30) {
	           printf("Error from write: %d, %d\n", wlen, errno);
	       }
	       tcdrain(fd);    /* delay for output */
	       usleep(4000*1000);
	       puts("Reading");
	       do {
	               unsigned char buf[30];
	               int rdlen;

	               rdlen = read(fd, buf, sizeof(buf) - 1);
	               if (rdlen > 0) {
  	                   unsigned char   *p;
	                   printf("Read %d:", rdlen);
	                   for (p = buf; rdlen-- > 0; p++)
	                       printf(" 0x%x", *p);
	                   printf("\n");
	               } else if (rdlen < 0) {
	                   printf("Error from read: %d: %s\n", rdlen, strerror(errno));
	               } else {  /* rdlen == 0 */
	                   printf("Timeout from read\n");
	                   break;
	               }
	               /* repeat read to get full message */
	           } while (1);
	puts("Done"); /* prints !!!Hello World!!! */
	return EXIT_SUCCESS;
}``

**Comment by AndriiSlyva** [comment]
Hello! I have three HiSense WiFi Air Conditioners, models AEH-W4E1.
I ran the air conditioning control server as an add-on to HomeAssistant.
![image](https://user-images.githubusercontent.com/116375817/197713760-5b1ad8ef-4544-44c7-a497-7ad1cd80460a.png)
But they are inactive
![image](https://user-images.githubusercontent.com/116375817/197714268-e0d39b3e-5f8d-46a1-9c4b-62a3e41cf6df.png)
What am I doing wrong? Help please!

**Comment by 1iverea9er** [comment]
@miky2k @straga 
Hi,
Can you explain in more detail the procedure of how you established a direct connection with AC?
Your link is broken ([https://pastebin.com/YYz2JiQY](https://github.com/deiger/AirCon/issues/url)).
I'm not a programmer. 
Can I do the same as you with esphome and integration [Modbus Controller ](https://esphome.io/components/modbus_controller.html) ?

**Comment by miky2k** [comment]
Link is working i try to fix , copy and paste if not work with click.
I'use https://github.com/jeelabs/esp-link  on esp8266  firmware to simulate com port   through wifi .

**Comment by straga** [comment]
> @miky2k 
I am store all about XM hisense.
https://github.com/straga/scrivo_project/tree/master/project/ac_xm_hisense_control/board_psram/root/module
https://github.com/straga/scrivo_project/blob/master/project/ac_xm_hisense_control/board_psram/root/module/scrivo_ac_xm/responses.py

**Comment by unecc** [comment]
Hi, some solution for hisense ac in hass?

**Comment by drak202** [comment]
I use the AEH-W4A1 module. Control and status is functional but unstable. Often the module disconnects from wifi regularly every 2 minutes. Regularly every 5 minutes I am unavailable in HA. I have already replaced the module with the new problem is the same. I assume that the problem is with the program in the module

**Comment by zoffypal** [comment]
I use https://github.com/pslawinski/esphome_airconintl and make some changes and it works with 1 model of my hisense AC. Now I can control AC via esphome but still can't getting any current status from AC, and this uart data not working for another model hisense AC.
Here's my repository https://github.com/zoffypal/esphome_hisence_ac

**Comment by zoffypal** [comment]
> @zoffypal HI Try good [micropython](https://github.com/straga/scrivo_project) or [ESPHome](https://github.com/straga/scrivo_project/issues/1) implementations. It's tested on AMS-09UR4SVEDB65 and works ![image](https://private-user-images.githubusercontent.com/82063500/336874034-a248ee9c-3be9-40e2-8fc5-5e1711d6855b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc5NDA5NTAsIm5iZiI6MTcxNzk0MDY1MCwicGF0aCI6Ii84MjA2MzUwMC8zMzY4NzQwMzQtYTI0OGVlOWMtM2JlOS00MGUyLThmYzUtNWUxNzExZDY4NTViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA5VDEzNDQxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiN2EwMWMyMTM2YWY2OTdkNGU3NzhmMGI0OWMxYzAxZWUxZWM5N2JlYzg5NTkwMGVlOGRlMmMxMjYzYWZkMGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JGhd2ABNXFYDw2IV8KDaqroeeUGoFynnbPZkhplqLz4) ![image](https://private-user-images.githubusercontent.com/82063500/336874100-50e6c097-6456-4c6c-920b-6f3991e29da1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc5NDA5NTAsIm5iZiI6MTcxNzk0MDY1MCwicGF0aCI6Ii84MjA2MzUwMC8zMzY4NzQxMDAtNTBlNmMwOTctNjQ1Ni00YzZjLTkyMGItNmYzOTkxZTI5ZGExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA5VDEzNDQxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5OTAzODZmMzc5YjFiMmY0MWU5MmQ2M2NlY2U2YWM0ODhhZDBhOTdmZmEyOGNlNGQyZGU1MzMyOTMzMWIwMjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Z9e64A6oJog6sPOZ67DfPxOAdLONNGT9cQcZOcu3zEY) ![photo_2024-06-05_17-12-04](https://private-user-images.githubusercontent.com/82063500/336874754-390d5891-a35a-4ce9-a922-412a4ee48eb8.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc5NDA5NTAsIm5iZiI6MTcxNzk0MDY1MCwicGF0aCI6Ii84MjA2MzUwMC8zMzY4NzQ3NTQtMzkwZDU4OTEtYTM1YS00Y2U5LWE5MjItNDEyYTRlZTQ4ZWI4LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA5VDEzNDQxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRhYzBiMTYyYWFjODdkNjU5ZDkxNzVmYjA0ZThhNjk5NzViOWIwNjc3MWJiZDQwMWViNWNkNTc3ZDcyNGM2OWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OvGqdmvelds9oo9L3D9Wxt_E5gzf2lTRCaUs5qSRiJo)

I was found this repository before. However, it's looks not esphome or adruino project, I didn't know how to use it neither to compile it and flash to esp8266 and connect to Home Assistant.

**Comment by 1iverea9er** [comment]
@zoffypal 
repo  is  micropython project, yes. 
If you want do this in ESPHome look at second link
